In [2]:
import pandas as pd
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import confusion_matrix

In [8]:
data = pd.read_excel("wireless_mlpclassifier_name.xlsx")
data

[Low price (낮은 가격)]  [Noise cancelling (노이즈 캔슬링)]  \
0                     1                             4   
1                     3                             5   
2                     4                             3   
3                     3                             4   
4                     4                             5   
..                  ...                           ...   
81                    3                             4   
82                    6                             3   
83                    5                             6   
84                    4                             6   
85                    6                             2   

    [Battery time (배터리 시간)]  [Light weight (가벼운 무게)]  [Brand (브랜드)]  \
0                         3                        5              6   
1                         4                        6              2   
2                         5                        6              2   
3                         6                        5              1   
4                         2                        1              6   
..                      ...                      ...            ...   
81                        2                        5              6   
82                        1                        4              5   
83                        1                        2              4   
84                        3                        5              1   
85                        3                        5              1   

    [Sound Quality (음질)]          wish device Brand preference  \
0                      2  Airpod pro (에어팟 프로)       Apple (애플)   
1                      1     Buds pro (버즈 프로)     Samsung (삼성)   
2                      1     Buds pro (버즈 프로)     Samsung (삼성)   
3                      2     Buds pro (버즈 프로)     Samsung (삼성)   
4                      3  Airpod pro (에어팟 프로)       Apple (애플)   
..                   ...                  ...              ...   
81                     1     Buds pro (버즈 프로)         No .(없음)   
82                     2      SONY WF-1000XM4         No .(없음)   
83                     3      SONY WF-1000XM4         No .(없음)   
84                     2     Airpod 3 (에어팟 3)       Apple (애플)   
85                     4  Airpod pro (에어팟 프로)       Apple (애플)   

   Are you currently using wireless earphone?   current device  \
0                                     Yes (예)               디렘   
1                                     Yes (예)    Buds 2 (버즈 2)   
2                                     Yes (예)    Buds 2 (버즈 2)   
3                                    No (아니오)              NaN   
4                                     Yes (예)  Airpod 2 (에어팟2)   
..                                        ...              ...   
81                                    Yes (예)           QCY T5   
82                                    Yes (예)  Airpod 2 (에어팟2)   
83                                    Yes (예)  Airpod 1 (에어팟1)   
84                                    Yes (예)  Airpod 2 (에어팟2)   
85                                    Yes (예)  Airpod 2 (에어팟2)   

    current satisfaction  
0                    4.0  
1                    5.0  
2                    5.0  
3                    NaN  
4                    4.0  
..                   ...  
81                   3.0  
82                   2.0  
83                   3.0  
84                   4.0  
85                   4.0  

[86 rows x 11 columns]

In [9]:
scaler = StandardScaler()
data['current satisfaction'] = scaler.fit_transform(data['current satisfaction'])

ValueError: Expected 2D array, got 1D array instead:
array=[ 4.  5.  5. nan  4.  4.  5.  4.  4.  4.  5.  4.  4.  4.  4. nan  5.  5.
  4.  4.  4.  4.  5.  4.  5.  4.  4.  5.  4.  4.  4.  4. nan  5.  4.  5.
  4.  4.  5.  3.  4.  5.  4.  3.  4.  5.  4.  4.  4.  4.  4.  4.  3. nan
  4.  4. nan  3.  4.  4. nan  4.  5.  4.  3.  4.  4. nan  4.  3.  5.  4.
  3.  3.  4.  4.  5.  4.  5.  3. nan  3.  2.  3.  4.  4.].
Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.

In [21]:
cs = data['current satisfaction']
cs.array
cs.reshape(-1,1)

AttributeError: 'Series' object has no attribute 'reshape'

In [37]:
training_set, validation_set = train_test_split(data, test_size = 0.2, random_state = 21)

In [38]:
X_train = training_set.iloc[:,0:6].values
y_train = training_set.iloc[:, -1].values
X_val = validation_set.iloc[:,0:6].values
y_val = validation_set.iloc[:, -1].values

In [39]:
def accuracy(confusion_matrix):
    diagonal_sum = confusion_matrix.trace()
    sum_of_all_elements = confusion_matrix.sum()
    return diagonal_sum/sum_of_all_elements

In [40]:
classifier = MLPClassifier(max_iter= 300,solver = 'adam', random_state = 1)
'''hidden_layer_sizes = (150,100,50),\
                           
    >> 이거 없애면 accuracy = 0.2777777777777778'''

'hidden_layer_sizes = (150,100,50),                           \n    >> 이거 없애면 accuracy = 0.2777777777777778'

In [41]:
classifier.fit(X_train, y_train)

C:\Users\jiwon\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:614: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (300) reached and the optimization hasn't converged yet.
  warnings.warn(


MLPClassifier(max_iter=300, random_state=1)

In [42]:
y_pred = classifier.predict(X_val)

In [43]:
cm = confusion_matrix(y_pred, y_val)

In [44]:
accuracy(cm)

0.3333333333333333

scaler 없애면 accuracy 0.2777777777777778

hidden layer도 디폴트면 0.3333333333333333

다른것들은 디폴트로 돌려도 정확도 변화 없음.